base/inference.jl

```diff
@@ -191,7 +193,7 @@ mutable struct InferenceState
                     vararg_type = rewrap(vararg_type, linfo.specTypes)
                 end
                 s_types[1][la] = VarState(vararg_type, false)
-                src.slottypes[la] = widenconst(vararg_type)
+                src.slottypes[la] = vararg_type
```

Comment:
I found it: `exprtype` reads `slottypes`, and inlining and some other passes call `exprtype`. It takes a really obscure case to trigger it. For example `f(x) = x()`, where `f` is inferred on the type of a singleton pure function. `x()` won't get inlined in that case because it's not a `Const`.